src/huggingface_hub/serialization/_dduf.py

```diff
@@ -0,0 +1,142 @@
+        """
+        with self.dduf_path.open("rb") as f:
+            f.seek(self.offset)
+            return f.read(self.length).decode(encoding=encoding)
+
+
+def read_dduf_file(dduf_path: Union[Path, str]) -> Dict[str, DDUFEntry]:
+    """
+    Read a DDUF file and return a dictionary of entries.
+
+    Only the metadata is read, the data is not loaded in memory.
+
+    Args:
+        dduf_path (`str` or `Path`):
+            The path to the DDUF file to read.
+
+    Returns:
+        `Dict[str, DDUFEntry]`:
+            A dictionary of [`DDUFEntry`] indexed by filename.
+
+    Raises:
+        - [`DDUFCorruptedFileError`]: If the DDUF file is corrupted (i.e. doesn't follow the DDUF format).
+    """
+    entries = {}
+    dduf_path = Path(dduf_path)
+    logger.info("Reading DDUF file %s", dduf_path)
+    with zipfile.ZipFile(str(dduf_path), "r") as zf:
+        for info in zf.infolist():
+            logger.debug("Reading entry %s", info.filename)
+            if info.compress_type != zipfile.ZIP_STORED:
+                raise DDUFCorruptedFileError("Data must not be compressed in DDUF file.")
+
+            # Use private attribute to get data range for this file.
+            # Let's reconsider later if it's too problematic (worse case, we can build our own metadata parser).
+            # Note: simply doing `info.header_offset + len(info.FileHeader())` doesn't work because of the ZIP64 extension.
+            offset = info._end_offset - info.compress_size  # type: ignore[attr-defined]
+
+            entries[info.filename] = DDUFEntry(
+                filename=info.filename, offset=offset, length=info.file_size, dduf_path=dduf_path
+            )
+    logger.info("Done reading DDUF file %s. Found %d entries", dduf_path, len(entries))
+    return entries
+
+
+def write_dduf_file(dduf_path: Union[str, Path], diffuser_path: Union[str, Path]) -> None:
+    """
+    Write a DDUF file from a diffusers folder.
+
+    A DDUF file is simply a ZIP archive with a few constraints (force ZIP64, no compression, only certain files).
+
+    Args:
+        dduf_path (`str` or `Path`):
+            The path to the DDUF file to write.
+        diffuser_path (`str` or `Path`):
+            The path to the folder containing the diffusers model.
+    """
+    # TODO: update method signature.
+    #       DDUF filename should be inferred as much as possible from high-level info (precision, model, etc.) to ensure consistency.
+    #       Example: "stable-diffusion-3.5-Q4-BNB.dduf"
+    #       See https://github.com/huggingface/diffusers/pull/10037#discussion_r1862275730.
+    logger.info("Writing DDUF file %s from folder %s", dduf_path, diffuser_path)
+    diffuser_path = Path(diffuser_path)
+    with zipfile.ZipFile(str(dduf_path), "w", zipfile.ZIP_STORED) as archive:
+        for path in diffuser_path.glob("**/*"):
+            if path.is_dir():
+                logger.debug("Skipping directory %s", path)
+                continue
```

Comment:
Why are we skipping directories? A `diffusers` formatted pipeline can have directories:
https://huggingface.co/black-forest-labs/FLUX.1-dev/tree/main

Or am I misunderstanding this?

Reply:
Ah thanks for explaining